src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+
+            List<Token> tokens = new();
+
+            if (addBeginOfSentence)
+            {
+                tokens.Add(new Token(BeginningOfSentenceId, BeginningOfSentenceToken, (0, 0)));
+            }
+
+            for (int index = 0; index != -1 && index < symbols.Length; index = symbols[index].next)
+            {
+                int id = symbols[index].id;
+                byte type = symbols[index].type;
+
+                if (id == UninitializedId)
+                {
+                    if (_vocab.TryGetValue(text.AsSpan().Slice(symbols[index].pieceSpan.Index, symbols[index].pieceSpan.Length), out (int Id, float Score, byte Type) tokenInfo))
+                    {
+                        id = tokenInfo.Id;
+                        type = tokenInfo.Type;
+                    }
+                    else
+                    {
+                        id = UnknownId;
+                        type = 0;
+                    }
+                }
+
+                if (type != (byte)ModelProto.Types.SentencePiece.Types.Type.Unused)
+                {
+                    if (id == UnknownId && ByteFallback)
+                    {
+                        EncodeAsBytes(text.AsSpan().Slice(symbols[index].pieceSpan.Index, symbols[index].pieceSpan.Length), symbols[index].pieceSpan.Index);
+                    }
+                    else
+                    {
+                        tokens.Add(new Token(
+                                    id,
+                                    GetTokenString(id, symbols[index].pieceSpan.Index, symbols[index].pieceSpan.Length, text.AsSpan()),
+                                    (symbols[index].pieceSpan.Index, symbols[index].pieceSpan.Length)));
+                    }
+                    continue;
+                }
+
+                Segment(symbols[index].pieceSpan, text.AsSpan());
+            }
+
+            ArrayPool<BpeSymbol>.Shared.Return(symbols);
+
+            if (addEndOfSentence)
+            {
+                tokens.Add(new Token(EndOfSentenceId, EndOfSentenceToken, (text.Length, 0)));
+            }
+
+            return tokens;
+
+            // Encode the Unknown token to bytes.
+            void EncodeAsBytes(ReadOnlySpan<char> text, int index)
+            {
+                for (int i = 0; i < text.Length; i++)
+                {
+                    char c = text[i];
+                    if (c <= 0x7F)
+                    {
+                        int id = (int)c + _byteCodeToIdOffset; // byte code is mapped to the to the Ids starting from 4.
+
+                        if (_vocabReverse.TryGetValue(id, out string? token))
+                        {
+                            tokens.Add(new Token(id, token, (index + i, 1)));
+                        }
+                    }
+                    else
+                    {
+                        Span<byte> utf8Bytes = stackalloc byte[100];
```

Comment:
Generally it's not good to have a stackalloc inside a loop. In this case, however, it looks like this will only ever be hit once because we always exit the loop after this code runs?

Reply:
That is right. When using the stackalloc in this place we exit the loop as we handle the rest of bytes all together. I'll keep it as it is to be closer where it is used inside the code.